rfcs/010-gateway-refactor/010-gateway-refactor.md

```diff
@@ -0,0 +1,205 @@
+tighten security the gateway can accept Keycloak access token only with the
+`ui-server` client as audience and reject all others.
+
+## Drawbacks
+
+- More traffic will go through the gateway
+- The gateway will have sessions with multiple clients
+
+## Rationale and Alternatives
+
+1. Why can't we simply use JWT access tokens everywhere?
+JWT access tokens from Keycloak expire and it is unsafe to accept expired
+access tokens. So if we go this route then everyone who is authenticated
+is forced to refresh their access token. But in order to refresh an access token
+the services need to receive additional credentials. In addition the services
+now much periodically check and worry about keeping their tokens refreshed. This
+should not be the responsibilities of renku components but rather on the gateway.
+The only expectation of Renku components we should have is that they verify a provided
+access token and confirm i) it is valid and ii) has the expected audience and/or claims.
+
+2. Why can't we use the same session ID for users and servers?
+The main reason is that these are expected to have different lifetimes.
+When a user logs out then their session is removed. But they may leave servers running
+that require access to credentials so that they continue operating properly (even
+when their creator has logged out). The server session with the gateway will be 
+removed when the server is shut down.
+
+3. Why not simply require a user/service to present a Keycloak user ID for authentication
+instead of the randomly generated session ID?
+Using the never-changing Keycloak user IDs for authentication is dangerous. This means
+that the moment someone finds out the ID for a user they can impersonate them. Whereas
+using temporary randomly generated session IDs is safer and even if the ID is stolen
+the user can be impersonated only for the duration of the session. Renku admins also
+have control over these sessions and they can force a specific user to be logged out. Whereas
+with static Keycloak user IDs the admin would have to fully remove or disable the user in
+Keycloak in order to log them out.
+
+4. Aren't we overlapping with resposibilities and functionality with the ui-server
+which already has a session with the user? 
+Yes we are. But I am not sure how much of a problem this is. The functionality described here
+has to be implemented somewhere and the ui-server does not provide this currently.
+
+## Unresolved questions
+
+Issues to be resolved through the RFC process before this is merged:
+- agreement on the resposiblities of the gateway
+- agreement on the design and decisions
+- whether to use more persistent storage for some sessions that need to be
+longer-lived (i.e. sessions for Jupyter servers)
+
+Issues to resolve through the implementation of this RFC:
+- potential elimination of traefik in the gateway (all golang web frameworks support reverse proxying out of the box)
```

Comment:
oh i was reading this thinking that it assumes Traefik is gone. Which adds significantly to the complexity of the implementation because now we have to also re-implement all the logic that is inside the middleware definition - not a bad thing because it's notoriously difficult to understand what is going on in there. otoh is implementing yet another proxy really something we want to be doing?

Reply:
IMO we use traefik only for very simple path rewriting and cookie stripping. So doing this in the golang proxy is really not hard. The golang server framework I would like to use (https://echo.labstack.com/) has very easy to use middlewares that give you much more flexibility and expressiveness than the traefik configuration. In addition we can also write unit tests for these whereas we cannot unit test the traefik configuration.

Lastly when we want sticky sessions we have to selectively remove some cookies in the gateway (but leave the sticky session cookie in there). Currently we remove all of them. To keep some cookies in a request but remove all others we need to use a custom plugin for traefik. I wrote one for this purpose [here](https://github.com/SwissDataScienceCenter/cookiefilter). I imagine that we may find ourselves in similar situations down the road as well where the functionality we need is not provided by traefik and we will have to write our own plugins. If we use the golang server framework to do this proxying then we do not have to worry about this.